### PR TITLE
Flaky blobs caps systemtest

### PR DESCRIPTION
**Pull request by eccles**
Reading the logs emitted by Debug log statements shows:

     2024-08-08T12:50:50.028Z    DEBUG    azblob/list.go:19    Count
     2024-08-08T12:50:50.032Z    DEBUG    azblob/list.go:36    Count 2

indicating that the code works and incurs 40ms delay only.

All systemtests pass - unfortunately the flaky blobs caps tests still fail

(the debug log statements have been removed)

[AB#5311](https://dev.azure.com/jitsuin/0629f48c-3979-4bbc-9026-cb06b3dfd0ae/_workitems/edit/5311)

### CODE REVIEW

File: azblob/list.go [hunk 1]
```diff
@@ -16,12 +16,23 @@ func (azp *Storer) Count(ctx context.Context, tagsFilter string, opts ...Option)
...
+		if err != nil {
+			return 0, err
+		}
+		count += int64(len(r.Items))
+		logger.Sugar.Debugf("Count %d (%v)", count, r.Marker)
+		if r.Marker == nil {
+			break
+		}
+	}
+	logger.Sugar.Debugf("Count %d", count)
```
Comment (robinbryce) [inline azblob/list.go:34]: Do we really need the log statement here ? I think it is better to leave this sort of telementry to the caller

File: azblob/list.go [hunk 2]
```diff
@@ -16,12 +16,23 @@ func (azp *Storer) Count(ctx context.Context, tagsFilter string, opts ...Option)
...
+	opts = append(opts, WithListMarker(&m))
+
+	for {
+		r, err := azp.FilteredList(ctx, tagsFilter, opts...)
+		if err != nil {
+			return 0, err
+		}
+		count += int64(len(r.Items))
+		logger.Sugar.Debugf("Count %d (%v)", count, r.Marker)
+		if r.Marker == nil {
```
Comment (robinbryce) [inline azblob/list.go:30]: I think this may need to be `r.Marker == nil || *r.Marker == ""` have a look at the other implementations to see what they do

Comment (eccles) [reply azblob/list.go]: I made the fix but it did not help. Serhiy has said the blobs caps limit tests must be rewritten...

File: azblob/list.go [hunk 3]
```diff
@@ -7,21 +7,27 @@ import (
...
+
+	for {
+		r, err := azp.FilteredList(ctx, tagsFilter, opts...)
+		if err != nil {
+			return 0, err
+		}
+		count += int64(len(r.Items))
+		if r.Marker == nil || *r.Marker == "" {
+			break
+		}
```
Comment (robinbryce) [inline azblob/list.go:28]: don't you need to asign 'm' here ?

File: azblob/list.go (hunk 2 above)
Comment (j-hartley) [reply azblob/list.go]: I am pretty sure you do need to check for marker is ""?  I found that problem elsewhere.

File: azblob/list.go [hunk 4]
```diff
@@ -68,7 +84,7 @@ func (azp *Storer) FilteredList(ctx context.Context, tagsFilter string, opts ...
 	}
 
 	if options.listMarker != nil {
-		span.SetTag("marker", *options.listMarker)
+		span.SetTag("marker", options.listMarker)
```
Comment (robinbryce) [inline azblob/list.go:87]: Why did you change this to set the ptr value ? We know it is not nil, we want to know its value don't we ?

File: azblob/list.go [hunk 5]
```diff
@@ -92,12 +108,11 @@ func (azp *Storer) FilteredList(ctx context.Context, tagsFilter string, opts ...
 		Items:      resp.Blobs,
 	}
 
-	r.Marker = resp.NextMarker
```
Comment (robinbryce) [inline azblob/list.go:95]: Why did you remove this line ? I think this completely breaks paging doesn't it ?

**Review by robinbryce — CHANGES_REQUESTED** [review]
I still don't understand your changes to FilteredList

File: azblob/list.go (hunk 2 above)
Comment (eccles) [reply azblob/list.go]: old code

File: azblob/list.go (hunk 4 above)
Comment (eccles) [reply azblob/list.go:87]: mistake - fixed

File: azblob/list.go (hunk 5 above)
Comment (eccles) [reply azblob/list.go:95]: its superfluous - r.Marker already has that value

File: azblob/list.go [hunk 6]
```diff
@@ -13,15 +13,25 @@ import (
...
-
-	return int64(len(r.Items)), nil
+	log := logger.Sugar.FromContext(ctx)
+	defer log.Close()
+
+	var count int64
+	var m ListMarker
+
+	for {
+		opts = append(opts, WithListMarker(m))
```
Comment (robinbryce) [inline azblob/list.go:23]: doesn't this opts list grow infinitely now ?

Comment (robinbryce) [reply azblob/list.go:23]: this needs to be 

r, err := azp.FilteredList(ctx, tagsFilter, append(opts,  WithListMarker(m))...)

**Review by robinbryce — CHANGES_REQUESTED** [review]
the opts in count grows redundantly and infinitely (at least to the number of pages)

**Review by robinbryce — APPROVED** [review]
LGTM

**Review by robinbryce — APPROVED** [review]
LGTM
